### PR TITLE
funccount: avoid map.zero and improve perf

### LINKED ISSUE #665: funccount "SyS_.*" stuck on-CPU
When run python funccount.py "SyS_.*",  at the beginning, log:
```
write(/sys/kernel/debug/tracing/kprobe_events, "p:kprobes/p_SyS_sigsuspend SyS_sigsuspend") failed: Device or resource busy
```
when Ctrl-C, the tracing not end:
```
Tracing 337 functions for "SyS_.*"... Hit Ctrl-C to end.
^C
ADDR             FUNC                          COUNT
ffffffff81097230 sys_getrusage                     1
ffffffff81253d10 SyS_fstatfs                       1
ffffffff81092cc0 sys_kill                          1
ffffffff81095170 sys_getpriority                   1
ffffffff8121d030 sys_creat                         1
ffffffff8108bd30 sys_capset                        1
ffffffff81238c00 sys_getcwd                        1
ffffffff8170d4c0 sys_getpeername                   1
ffffffff81267d20 sys_epoll_create                  1
ffffffff81228220 SyS_pipe                          2
ffffffff810af5e0 sys_sched_getaffinity             2
ffffffff8170d6f0 SyS_shutdown                      2
ffffffff8121c2f0 sys_chdir                         2
ffffffff8113f250 sys_seccomp                       2
ffffffff8122fc60 sys_rmdir                         2
ffffffff8121d010 sys_openat                        2
ffffffff81223db0 SyS_readlink                      2
ffffffff810808f0 sys_set_tid_address               3
ffffffff8122fb60 sys_mkdir                         3
ffffffff81097140 sys_setrlimit                     3
ffffffff81228170 SyS_pipe2                         3
ffffffff812269e0 SyS_execve                        3
ffffffff8170da10 sys_sendmmsg                      3
ffffffff8170d470 sys_accept4                       3
ffffffff81097280 sys_umask                         3
ffffffff8123d620 sys_dup2                          4
ffffffff81085f50 sys_exit_group                    5
ffffffff81234d60 sys_ppoll                         6
ffffffff8170d480 sys_accept                        6
ffffffff8126bd10 sys_eventfd2                      7
ffffffff81086120 sys_wait4                         9
ffffffff8170d3b0 sys_bind                          9
ffffffff8122fcb0 sys_unlink                       10
ffffffff81096590 sys_newuname                     12
ffffffff81230d30 sys_rename                       12
ffffffff8170d4b0 sys_getsockname                  13
ffffffff81094f20 sys_setpriority                  13
ffffffff811d1d60 sys_brk                          15
ffffffff81096ec0 sys_getrlimit                    15
ffffffff8123d730 sys_dup                          16
ffffffff8151db10 SyS_getrandom                    17
ffffffff81085e80 SyS_exit                         18
ffffffff8170d4a0 sys_connect                      22
ffffffff81252890 sys_fdatasync                    23
ffffffff8126b0e0 sys_timerfd_settime              24
ffffffff810972b0 sys_prctl                        25
ffffffff81252870 sys_fsync                        26
ffffffff8121f940 sys_pread64                      27
ffffffff8121df40 sys_lseek                        27
ffffffff810923b0 sys_rt_sigprocmask               30
ffffffff8170d160 sys_socketpair                   31
ffffffff8170d610 sys_getsockopt                   32
ffffffff81085fa0 sys_waitid                       32
ffffffff8170d530 sys_setsockopt                   32
ffffffff81232ad0 sys_ioctl                        33
ffffffff81080e00 sys_clone                        35
ffffffff8170d0b0 sys_socket                       35
ffffffff81104fd0 SyS_set_robust_list              36
ffffffff8121bdf0 sys_ftruncate                    36
ffffffff81223c70 sys_newfstatat                   37
ffffffff810f2ae0 sys_nanosleep                    54
ffffffff810ec3e0 sys_kcmp                         58
ffffffff81223c90 sys_readlinkat                  118
ffffffff810f4a00 sys_clock_gettime               122
ffffffff81267df0 SyS_epoll_ctl                   123
ffffffff81234410 SyS_pselect6                    131
ffffffff8121f9f0 sys_pwrite64                    139
ffffffff8170d4d0 SyS_sendto                      182
ffffffff81253cf0 sys_statfs                      235
ffffffff811d3c00 SyS_mprotect                    267
ffffffff812646c0 sys_inotify_add_watch           441
ffffffff81093500 sys_rt_sigaction                514
ffffffff810f1460 SyS_alarm                       675
ffffffff8121c0c0 sys_access                      758
ffffffff81223c60 sys_newlstat                   1392
ffffffff811d0ff0 sys_munmap                     3126
ffffffff811d0450 sys_mmap_pgoff                 3410
ffffffff810341f0 sys_mmap                       3532
ffffffff812313f0 sys_fcntl                      3665
ffffffff81223c50 sys_newstat                   10342
ffffffff811ddd00 sys_madvise                   10452
ffffffff81233000 sys_getdents                  11831
ffffffff8170d810 sys_sendmsg                   13606
ffffffff81223c80 sys_newfstat                  27611
ffffffff8121cff0 sys_open                      33712
ffffffff8121b0e0 sys_close                     33929
ffffffff8121f880 sys_write                     46226
ffffffff81268c10 sys_epoll_wait                67218
ffffffff8121f7c0 sys_read                      89183
ffffffff81234300 sys_select                   161449
ffffffff8170d500 sys_recvfrom                 194010
ffffffff810f3380 sys_setitimer                297644
ffffffff81234c30 sys_poll                     407389
ffffffff8121fbd0 sys_writev                   483703
ffffffff8170dac0 sys_recvmsg                  730047
ffffffff81105df0 sys_futex                    932437
^C
^C
^C
^C
^C
^C
^C
^C
```
check the tracing pid cpu use:
```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
28300 root      20   0  118424  57928  31504 R 100.0  2.8   1:41.93 python
```
I need to kill -9 PID to end the process.

Find info about the problem, the commit [0d93605b468f6a54980a81fd72952a0f2ade42a4](https://github.com/iovisor/bcc/commit/0d93605b468f6a54980a81fd72952a0f2ade42a4) and issuse [#233](https://github.com/iovisor/bcc/issues/233)

bitsect the bcc, find commit reintroduced the deadlock problem:
```
commit 95728834ee60875b251a2696dd134f1c11d61979
Author: Brenden Blanco <bblanco@plumgrid.com>
Date:   Tue Feb 23 13:13:57 2016 -0800

    Use pre-calculated function addresses in funccount

    Hash map updates are not safe in recursive context. Change the
    implementation of funccount to be read-only by converting
    lookup_or_init() to lookup(). When run in interval mode, zero the counts
    table instead of clearing it.

    Signed-off-by: Brenden Blanco <bblanco@plumgrid.com>

diff --git a/tools/funccount.py b/tools/funccount.py
index f417087..427e7c1 100755
--- a/tools/funccount.py
+++ b/tools/funccount.py
@@ -65,9 +65,13 @@ BPF_HASH(counts, struct key_t);
 int trace_count(struct pt_regs *ctx) {
     FILTER
     struct key_t key = {};
-    u64 zero = 0, *val;
-    key.ip = ctx->ip;
-    val = counts.lookup_or_init(&key, &zero);
+    u64 *val;
+    // the kprobe pc is slightly after the function starting address, align
+    // back to the start (4 byte alignment) in order to match /proc/kallsyms
+    key.ip = ctx->ip & ~3ull;
+    val = counts.lookup(&key);
+    if (!val)
+        return 0;
     (*val)++;
     return 0;
 }
@@ -81,6 +85,16 @@ else:
 if debug:
     print(bpf_text)
 b = BPF(text=bpf_text)
+counts = b.get_table("counts")
+
+# pre-insert the function addresses into the counts table
+fns = b._get_kprobe_functions(pattern)
+for fn in fns:
+    addr = b.ksymname(fn)
+    if addr == -1:
+        raise Exception("Unknown symbol name %s" % fn)
+    counts[counts.Key(addr)] = counts.Leaf()
+
 b.attach_kprobe(event_re=pattern, fn_name="trace_count")
 matched = b.num_open_kprobes()
 if matched == 0:
@@ -106,10 +120,10 @@ while (1):
         print("%-8s\n" % strftime("%H:%M:%S"), end="")

     print("%-16s %-26s %8s" % ("ADDR", "FUNC", "COUNT"))
-    counts = b.get_table("counts")
     for k, v in sorted(counts.items(), key=lambda counts: counts[1].value):
+        if v.value == 0: continue
         print("%-16x %-26s %8d" % (k.ip, b.ksym(k.ip), v.value))
-    counts.clear()
+    counts.zero()

     if exiting:
         print("Detaching...")

```
debug funccount program, this line change introduce  the problem:
```Python
-    counts.clear()
+    counts.zero()
```
back to counts.clear(), not reproduce the deadlock problem. if I change back to counts.clear(), the cache of function addresses don't work. 

Step to step into funccount, cause by table.py：
```Python
    def __setitem__(self, key, leaf):
        key_p = ct.pointer(key)
        leaf_p = ct.pointer(leaf)
        res = lib.bpf_update_elem(self.map_fd,
                ct.cast(key_p, ct.c_void_p),
                ct.cast(leaf_p, ct.c_void_p), 0)
        if res < 0:
            errstr = os.strerror(ct.get_errno())
            raise Exception("Could not update table: %s" % errstr)
```
guess the code race bpf program.

Thanks.

### PR DESCRIPTION
**Pull request by brendangregg**
This fixes #665, by modifying funccount to leave the function counts as cumulative and never zero them, instead, calculating deltas for the interval option. Eg, try running:

```
./funccount.py -i 1 'bash:r*'
```

before and after to see the fix.

I'm not sure if it should be fixed in funccount or in bcc, but this funccount change seems straightforward. CC @goldshtn

### CODE REVIEW

**Comment by goldshtn** [comment]
Would it help if we didn't prepopulate the table and then just call `clear()` from the Python side? It just means the probes need to call `lookup_or_init` instead of `lookup`, not that big a deal probably. And then this workaround won't be necessary, I think. @brendangregg

**Comment by brendangregg** [comment]
Switching from clear() to zero() was necessary several kernel versions ago to avoid an issue; I assumed it was still necessary, I'll have to check...

File: tools/funccount.py [hunk 1]
```diff
@@ -262,12 +263,20 @@ def run(self):
             print("%-36s %8s" % ("FUNC", "COUNT"))
             counts = self.probe.bpf["counts"]
             for k, v in sorted(counts.items(),
-                               key=lambda counts: counts[1].value):
+                               key=lambda counts: counts[1].value -
+                               (self.prev[counts[0].value] if
+                               counts[0].value in self.prev else 0)):
```
Comment (markdrayton) [inline tools/funccount.py]: `self.prev.get(counts[0].value, 0)` is a little more concise.

File: tools/funccount.py [hunk 2]
```diff
@@ -262,12 +263,20 @@ def run(self):
...
                 if v.value == 0:
                     continue
-                print("%-36s %8d" %
-                      (self.probe.trace_functions[k.value], v.value))
-            counts.zero()
+                val = v.value
+                # for intervals, calculate delta from prev[] measurement
+                if self.args.interval:
+                    if k.value in self.prev:
+                        val = val - self.prev[k.value]
```
Comment (markdrayton) [inline tools/funccount.py]: Here too — `val = val - self.prev.get(k.value, 0)`

**Comment by markdrayton** [comment]
Switching to not prepopulating would also allow optionally breaking down the number of function calls by PID and/or TID, which is something I often want to do. (I don't recall there being another tool which does this but please point me at it if there is!)

**Comment by goldshtn** [comment]
In #780 I propose the fix whereby we don't prepopulate the map anymore (and also fixed a bug in the Python module). It worked in my limited testing.

**Comment by brendangregg** [comment]
thanks @markdrayton @goldshtn ; I'll close this since the better fix is in #780

**Pull request closed without merging**